core/domain/acl_decorators_test.py

```diff
@@ -206,43 +216,46 @@ def setUp(self):
         rights_manager.publish_collection(self.owner, self.published_col_id)
 
     def test_guest_is_redirected_to_login_page(self):
-        response = self.testapp.get(
-            '/mock/%s' % self.published_col_id, expect_errors=True)
+        with self.swap(self, 'testapp', self.mock_testapp):
```

Comment:
You only need to swap when calling get_json. Here there is no need to swap. (ditto at other 302 responses)

Also just to note, @seanlip only 302 responses cannot be tested using get_json because it's a redirect and returns a html. So for now I think let's test 302 by directly calling self.testapp.get (or mock_testapp).

Reply:
I don't think I understand (conceptually, at least). If we make a GET JSON request, and it redirects, it seems to me that it should redirect to a JSON handler. Unless it redirects because the user doesn't have permission or something (maybe they aren't logged in)? Then in such cases I guess we should 404...

If this is the case, can we update the "check logged in" decorator to do different things based on whether the handler has json return type or not?

Reply:
@nithusha21, I found this:
here it says it will raise 401
https://github.com/oppia/oppia/blob/f11089bcb4c5fc0bd4717c85224c465796148ed0/core/controllers/base.py#L102-L103

but here it says it will send a redirect uri
https://github.com/oppia/oppia/blob/f11089bcb4c5fc0bd4717c85224c465796148ed0/core/controllers/base.py#L447-L450